### PR TITLE
[WIP] Support common block in OpenMP Private Clause

### PR DESCRIPTION
**Pull request by PeixinQiao**
This patch adds some missed test cases and supports common block in OpenMP Private Clause. Also remove some unused code.

### CODE REVIEW

**Comment by jeanPerier** [comment]
Two high level questions here:

1. Are their any OpenMP requirements specific to common blocks in private or is this completely equivalent to specifying all the common block members in the private statement (e.g, do we have to guarantee a single continuous allocation block...) ?

2. Are symbols with HostAssocDetails created for the members of the common block appearing in a Common block ?

If 2. holds true, I think it would be much cleaner if you were to associate the newly created extended values for the private members to the symbols with HostAssocDetails, preventing the ad-hoc handling in gen(Symbol). Assuming you have the OpenMP construct scope at hand, you can simply get the symbols with HostAssocDetails by looking for the symbols with the common block member names (name of the `mem` symbols in your code) in this scope, and then bind that.
If they are not any specific requirements to CommonBlock, you could just use the previous createHostAssociateVarClone once you have these symbols (but if you need a single allocation, and you may need to, then you still need to do something specific here).

**Comment by PeixinQiao** [comment]
> * Are their any OpenMP requirements specific to common blocks in private or is this completely equivalent to specifying all the common block members in the private statement (e.g, do we have to guarantee a single continuous allocation block...) ?

I don't see any specific requirements in OpenMP 5.1 Spec. I think it is completely equivalent to specifying all the common block members in the private statement. No need to guarantee a single contiguous allocation block.

> * Are symbols with HostAssocDetails created for the members of the common block appearing in a Common block ?

No. Thanks for the explanations. I think symbols in common block is missed when making HostAssoc symbols. Let me try to create HostAssocDetails  for common block members.

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -351,36 +338,72 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
+      auto allocaVal = allocateThreadLocal(
+          sym, genType(sym), fir::factory::getExtents(*builder, loc, hexv),
+          fir::getTypeParams(hexv));
+      auto exval = fir::substBase(hexv, allocaVal);
+      if (!bindSymbol(sym, exval))
+        return false;
+    } else {
+      return false;
+    }
+    return true;
```
Comment (schweitzpgi) [inline flang/lib/Lower/Bridge.cpp:406]: These `return`s are a bit of an eye-snag. https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

**Review by schweitzpgi — CHANGES_REQUESTED** [review]
Flagging this as having merge conflicts to be resolved.

**Comment by PeixinQiao** [comment]
> Flagging this as having merge conflicts to be resolved.

Thanks @schweitzpgi for the comments. I am waiting for https://reviews.llvm.org/D112876 to be merged and cherry-picked. It affects the implementation of this patch.

**Comment by PeixinQiao** [comment]
Abandon this currently since the original code is the same as that for threadprivate directive. Now the code for threadprivate is totally different from this. I don't have the time to look at this now.

**Pull request closed without merging**